### DISCUSSION TITLE
How to pass keystore details in runtime

### DISCUSSION THREAD

**Discussion by chamikaCN** [phax/phase4#88] [Q&A] [answered]
Hi Philip,

Is there a particular way to pass sensitive information such as key store passwords to the server examples you have implemented during runtime instead of having them in `application.properties` file?

I used the `AS4CryptoFactoryInMemoryKeyStore` for the same purpose in the client example. but cannot find a way to do that in server examples.

**Comment by phax** [comment, marked as answer]
Hi Chamika,
in general `AS4CryptoFactoryInMemoryKeyStore` is the tool of choice, to pass in "dynamic" keystore configurations.
There are ways to do something similar for receiving, but it very much depends on the logic that you require for choosing the correct certificate. If you could briefly elaborate on that logic by which you want to choose certificates, I can make more specific suggestions

**Reply by chamikaCN** [reply]
Thanks Philip,
I have successfully set up and run both `phase4-server-webapp` and `phase4-peppol-server-webapp` projects. due to security requirements, I tried to pass Keystore details dynamically in the `phase4-server-webapp` example. As you have specified it reads all keystore-related details from `application.properties` But was not able to find how to achieve that. I am currently using a similar sever to the one you have included in the tests. 
From where can I assign an `AS4CryptoFactoryInMemoryKeyStore` to be the default cryptofactory instead of `AS4CryptoFactoryProperties (AS4CryptoProperties.createFromConfig ())`?
Ideally a method tobe overridden or setup a property from `AS4WebAppListener`

**Reply by phax** [reply]
Ah okay. In that case you need to overwrite the `com.helger.phase4.servlet.AS4Servlet` and specify your own implementation of the AS4 Servlet.

The default constructor code of the servlet is:
```java
  // Multipart is handled specifically inside
    settings ().setMultipartEnabled (false);
    // HTTP POST only
    handlerRegistry ().registerHandler (EHttpMethod.POST, new AS4XServletHandler ());
```
Change it like this to provide the crypto factory from a different source:
```java
  // Multipart is handled specifically inside
    settings ().setMultipartEnabled (false);
    // HTTP POST only
    final AS4XServletHandler hdl = new AS4XServletHandler ();
    hdl.setCryptoFactorySupplier (() -> TODO);
    handlerRegistry ().registerHandler (EHttpMethod.POST, hdl);
```
hth

**Reply by chamikaCN** [reply]
Will try it out :slightly_smiling_face: 

Thanks Phillip

**Reply by chamikaCN** [reply]
It fulfilled my requirement, Thanks again
